### PR TITLE
Add automation to create a gitlab project

### PR DESCRIPTION
**Pull request by dave-shawley**
This PR builds on top of #72 by adding the `imbi.automations.gitlab.create_project` function that can be used as an automation callable. It does required that you have an OAuth 2 application defined in GitLab that is used to authorize Imbi to act as a specific GitLab user. You will need the OAuth 2 Client ID and Client Secret for that application to continue (details below).

Once you have the GitLab connection configured and a user connected, creating a new project of a project type that has automations enabled will add toggles for each available automation to the display.
image

After creating the Imbi project, the project creation endpoint will run the selected automations (8400ff4b2a686fb2df29b71fe2bffe00d3790888). If none of the automations fail, then the endpoint responds with a 👍 to the client. If any automation fails, then any successful automations are rolled back, the Imbi project is deleted, and a 👎 is returned to the client. The only automation implemented in this PR is one to create a GitLab project (e7757205f391aee0fbea06ec6839f4cacc349313). The project location is determined by the `namespace.gitlab_group_name` and `project_type.gitlab_project_prefix`. If either is missing, then the slug is substituted. There are a few interesting error conditions that can pop up:

1. since the GitLab creation request is done as the connected GitLab user, you might not have access to create projects in the target -- _this one is handled with a reasonable message IIRC_
2. if the target GitLab parent doesn't exist, a "configuration error" is displayed -- _made this decision since the namespace or project-type data in Imbi is misconfigured_
3. you can configure interesting combinations of GitLab group and sub-group permissions that could result in _really strange errors_

Feel free to play around with different settings. The error handling can be changed fairly easily... as long as we can determine what to do programatically.

You will also need https://github.com/AWeber-Imbi/imbi-ui/pull/73 running to see the new UI options.

## Guru meditations

* should we make the project defaults configurable under `automations:gitlab:` in the configuration file?
* is using a default of slugs appropriate instead of explicit configuration?
* should we send automation dependency information to the client?  Attempts to create a project with unmet dependencies will fail today but I didn't spend a lot of time thinking through alternatives.

### GitLab OAuth 2 Application Details
**Confidential** ☑️ \
**Scopes** `api` \
**Redirect URI** (one per line)
```
http://localhost:8000/gitlab/auth
http://127.0.0.1:8000/gitlab/auth
```

<details>
<summary><h3>Script to enable GitLab OAuth 2 integration & automation</h3></summary>

```python
#!/usr/bin/env python3

import contextlib
import getpass
import http.client
import json
import os
import sys
import textwrap
import urllib.parse


def get_string(prompt, secret=False):
    env_var = prompt.replace(' ', '_').upper()
    with contextlib.suppress(KeyError):
        return os.environ[env_var]

    try:
        if secret:
            return getpass.getpass(prompt + ': ')
        return input(prompt + ': ')
    except (IOError, KeyboardInterrupt):
        sys.exit('Canceled')


def post(conn, path, body, allow_statuses=None):
    allow_statuses = allow_statuses or []
    headers = {'Private-Token': imbi_api_token,
               'Content-Type': 'application/json'}
    conn.request('POST', path,
                 json.dumps(body).encode(),
                 headers=headers)
    rsp = conn.getresponse()
    rsp_body = rsp.read()
    if rsp.code >= 400:
        print('Imbi API Failure: POST {} -> {}'.format(
            path, rsp.code))
        if rsp.code in allow_statuses:
            print('Failure ignored here')
            print('')
        else:
            if rsp_body:
                json.dump(json.loads(rsp_body), sys.stdout, indent=2)
                print('')
            sys.exit(1)
    return json.loads(rsp_body) if rsp_body else None


if not os.environ.get('IMBI_API_TOKEN'):
    os.environ['IMBI_API_TOKEN'] =  '00000000-0000-0000-0000-000000000000'
imbi_api_token = get_string('Imbi API Token', secret=True)
gitlab_client_id = get_string('GitLab Client ID')
gitlab_client_secret = get_string('GitLab Client Secret', secret=True)
gitlab_url = get_string('GitLab URL')
imbi_url = get_string('Imbi URL')

parsed = urllib.parse.urlparse(gitlab_url)
gitlab_url = parsed.scheme + '://' + parsed.netloc

parsed = urllib.parse.urlparse(imbi_url)
if parsed.scheme == 'http':
    conn = http.client.HTTPConnection(parsed.hostname, parsed.port or 80)
elif parsed.scheme == 'https':
    conn = http.client.HTTPSConnection(parsed.hostname, parsed.port or 443)
else:
    sys.exit('Unhandled scheme: ' + parsed.scheme)
imbi_url = parsed.scheme + '://' + parsed.netloc

with contextlib.closing(conn):
    print('Created HTTP API project type')
    post(conn, '/project-types', {
        'name': 'HTTP API',
        'plural_name': 'HTTP APIs',
        'slug': 'api',
        'icon_class': 'fas cog',
        'environment_urls': True,
        'gitlab_project_prefix': 'api',
    }, allow_statuses=[409])

    print('Creating GitLab integration')
    post(conn, '/integrations', {
        'name': 'gitlab2',
        'api_endpoint': gitlab_url + '/api/v4',
        'api_secret': None,
    })

    print('Configuring GitLab OAuth2 connection')
    post(conn, '/integrations/gitlab2/oauth2', {
        'authorization_endpoint': gitlab_url + '/oauth/authorize',
        'token_endpoint': gitlab_url + '/oauth/token',
        'revoke_endpoint': None,
        'callback_url': imbi_url + '/gitlab/auth',
        'client_id': gitlab_client_id,
        'client_secret': gitlab_client_secret,
        'use_pkce': True,
    })

    print('Creating Create GitLab Project automation')
    post(conn, '/integrations/gitlab2/automations', {
        'name': 'Create GitLab Project',
        'callable': 'imbi.automations.gitlab.create_project',
        'categories': ['create-project'],
        'applies_to': ['api']
    })


term_size = os.get_terminal_size()

def display(*paragraphs):
    print()
    for p in paragraphs:
        msg = textwrap.dedent(p.format(**globals()))
        for line in textwrap.wrap(msg, term_size.columns * 0.7):
            print(line.strip())
        print()

display(
'''
    Your Imbi instance has been configured to connect to {gitlab_url}.
    Please open {imbi_url}/ui/user/profile and click the "Connect to
    GitLab" button.
''', '''
    If the button is not showing, reload the page. If that doesn't work
    then restart your Imbi instance.
'''
)
```
</details>

### CODE REVIEW

File: imbi/automations/__init__.py [hunk 1]
```diff
@@ -1,2 +1,206 @@
...
+
+    async def _run_cleanups(self, exc: BaseException) -> None:
+        cleanups, self._cleanups = self._cleanups, []
+        for cleanup in reversed(cleanups):
+            try:
+                maybe_coro = cleanup(self, exc)
+                if inspect.isawaitable(maybe_coro):
+                    await maybe_coro
+            except Exception as error:
+                self.logger.exception('cleanup %r failed with %s', cleanup,
```
Comment (in-op) [inline imbi/automations/__init__.py:148]: I wonder if we should let the user know that a cleanup has failed. Or maybe that's too remotely corner-case to care, and we just have them manually fix it.

Comment (dave-shawley) [reply imbi/automations/__init__.py:148]: I'd lean towards the latter at this stage. If we improve the feedback (eg, SSE) then we can add it there.

File: imbi/clients/gitlab.py [hunk 2]
```diff
@@ -222,3 +236,34 @@ async def fetch_user_namespace(self):
...
+        })
+        self.logger.info('refreshing token for %s (%s)', self.user.username,
+                         self.token.external_id)
+        self.logger.debug('refresh_request: %r', body)
+        response = await super().http_fetch(
+            str(self.token.integration.token_endpoint),
+            method='POST',
+            body=body,
+            content_type='application/x-www-form-urlencoded',
+        )
```
Comment (in-op) [inline imbi/clients/gitlab.py:279]: Can we do basic auth here? Similar to 
https://github.com/AWeber-Imbi/imbi-api/pull/68/commits/3ec775af97e5e35c58846be2d3bc5f8cd26ba844

Comment (dave-shawley) [reply imbi/clients/gitlab.py:279]: Lemme try that ... _they only supported IDs in request body in earlier versions_

**Pull request merged**